server/api/studio/pkg/client/session.go

```diff
@@ -0,0 +1,96 @@
+package client
+
+import (
+	"sync"
+	"time"
+
+	nebula "github.com/vesoft-inc/nebula-go/v3"
+	"github.com/zeromicro/go-zero/core/logx"
+)
+
+type SessionPool struct {
+	sessions map[int64]*ClientSession
+	mu       sync.Mutex
+	waitlist []ChannelRequest
+	waitMu   sync.Mutex
+}
+
+type ClientSession struct {
+	session  *nebula.Session
+	id       int64
+	isLocked bool
+	usedTime time.Time
+}
+
+func (client *Client) createClientSession() (session *ClientSession, id int64, err error) {
+	_session, err := client.graphClient.GetSession(client.account.username, client.account.password)
+	if err != nil {
+		return nil, id, err
+	}
+	id = _session.GetSessionID()
+	clientSession := ClientSession{session: _session, id: id}
+	return &clientSession, id, nil
+}
+
+func (client *Client) getSession() (session *ClientSession, err error) {
+	pool := client.sessionPool
+	pool.mu.Lock()
+	defer pool.mu.Unlock()
+	if len(pool.sessions) == 0 {
+		// init first session
+		clientSession, id, err := client.createClientSession()
+		if err != nil {
+			logx.Errorf("[Init first session failed]: %+v", err)
+			return nil, err
+		}
+		pool.sessions[id] = clientSession
+		clientSession.isLocked = true
+		return clientSession, nil
+	}
```

Comment:
These seem to be unnecessary? The following will deal with the situation where the session is not obtained.

Reply:
its not same. we need to return err if creating first session failed，but we can add request to wait list if creating other session failed，so it don't return err